### PR TITLE
Removed duplicate criteria, which seems to increase performance ~1000X

### PR DESCRIPTION
**Pull request by ee-usgs**
1000X performance improvement for queries that contain Country, State and County criteria

* Most queries that contain counties also contain state and country b/c of how the UI is built.
* Also added some logging, but its not perfect

Before making a pull request
----------------------------

- [ ] Make sure all tests run
- [ ] Update the changelog appropriately

Title
-----------
Brief description of changes. Reference the JIRA ticket if appropriate

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [ ] If appropriate, put the link to the PR in the JIRA ticket
- [ ] Assign someone to review unless the change is trivial

### CODE REVIEW

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+	void dedupParametersTwoCountryOnly() {
+
+		params.setCountrycode(Arrays.asList("US","CH"));
+		FilterParameters orgParams = fpClone(params);
+
+		FilterParametersUtil.dedupParameters(params);
+
+		assertEquals(2, params.getCountrycode().size());
+		assertThat(params.getCountrycode()).containsExactlyInAnyOrderElementsOf(orgParams.getCountrycode());
+		assertTrue(params.getStatecode().isEmpty());
```
Comment (orosz-usgs) [inline src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:71]: state code was empty from the beginning, could be a false positive.

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+		params.setCountrycode(Arrays.asList("US", "CH")); //Drop all - covered by county
+		params.setStatecode(Arrays.asList("US:06")); //Drop all - covered by county
+		params.setCountycode(Arrays.asList("US:06:047"));
+
+		FilterParameters orgParams = fpClone(params);
+
+		FilterParametersUtil.dedupParameters(params);
+
+		assertTrue(params.getCountrycode().isEmpty());
```
Comment (orosz-usgs) [inline src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:222]: Shouldn't the non  US country still be included ('CH' china).

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+		params.setCountrycode(Arrays.asList("US", "CH")); //Drop all - covered by county
+		params.setStatecode(Arrays.asList("US:06", "US:07")); //Drop all - covered by county
+		params.setCountycode(Arrays.asList("US:06:047", "US:06:048", "US:08:999"));	//US:08:999 should be dropped
+
+		FilterParameters orgParams = fpClone(params);
+
+		FilterParametersUtil.dedupParameters(params);
+
+		assertTrue(params.getCountrycode().isEmpty());
```
Comment (orosz-usgs) [inline src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:254]: shouldn't the second country (CH) china still be there?

**Review by orosz-usgs — COMMENTED** [review]
Awesome for putting these together very quickly! like you said there are probably corner cases,
we have not considered. Also, I could be misunderstanding WQP query search params.

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	void setUp() {
+		params = new FilterParameters();
+	}
+
+	@Test
+	void nonNullModifiableList() {
+		List orgList = Arrays.asList("US");
+		List newList = FilterParametersUtil.nonNullModifiableList(orgList);
+
+		assertTrue(orgList != newList);
```
[the remark this answers is not in the document]
Comment (ee-usgs) [reply src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:25]: This assertion is that the newList is not the same object as the original list.  I really want the simple '!=' in this case.

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+		assertThat(orgList).containsAnyElementsOf(newList);
+		newList.clear();	//should be OK
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:27]: Line 27 verifies that the newList is modifiable.
Line 26 is perhaps not the best test - it really should be the 'contains all the elements in any order' test, but since there is only one element, it works in this case.

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java (hunk 1 above)
Comment (ee-usgs) [reply src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:71]: True - I just have this pattern in each test of verifying the values of Country, State and County each time.  I could remove the ones that are unchanged, but I guess I prefer the consistency and clarity of stating exactly what I expect each time.

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java (hunk 2 above)
Comment (ee-usgs) [reply src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:222]: Nope - Its logically the same to only include the county and drop the countries and state - It will have the same result as the original.

You can test that out - convince yourself that this is true :-)

To help thinking about that, remember that within each tier (country, state and county) it is OR logic.
Between the tier's it is AND logic.
Country == CH AND County == "US:06:047" never matches any records.
Country == US AND County == "US:06:047" does match records.
CH can be dropped b/c it does not add any matching records to the query.
State can be dropped because the county is within it and is more specific.

File: src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java (hunk 3 above)
Comment (ee-usgs) [reply src/test/java/gov/usgs/wma/wqp/parameter/FilterParametersUtilTest.java:254]: Nope - See previous comments.

I do really need a solid review on the logic here, so please convince yourself that either "CH" can absolutely be dropped or it absolutely cannot and we need to make a change.

Try out some records with pencil and paper or some example queries if needed, but we need 100% confidence on this.
